src/dfx/src/commands/pull.rs

```diff
@@ -95,22 +114,138 @@ async fn fetch_deps_to_pull(
+    logger: &Logger,
+    canister_id: Principal,
+) -> DfxResult {
+    info!(logger, "Downloading wasm of canister {canister_id}...");
+
+    let agent = agent_env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    // try fetch `dfx:wasm_hash`. If not available, get the hash of the on chain canister.
+    let hash_on_chain = match fetch_metatdata(agent, canister_id, DFX_WASM_HASH).await {
+        Ok(Some(wasm_hash_raw)) => {
+            let wasm_hash_str = String::from_utf8(wasm_hash_raw)?;
+            info!(
+                logger,
+                "Canister {canister_id} specified a custom hash: {wasm_hash_str}"
+            );
+            hex::decode(wasm_hash_str)?
+        }
+        Ok(None) => {
+            let canister_status =
+                get_canister_status(agent_env, canister_id, &CallSender::SelectedId).await?;
+            match canister_status.module_hash {
+                Some(hash_on_chain) => hash_on_chain,
+                None => {
+                    bail!("Canister {canister_id} doesn't have module hash. Perhaps it's not installed.");
+                }
+            }
+        }
+        Err(e) => {
+            bail!(e);
+        }
+    };
+
+    // target $HOME/.cache/dfinity/wasms/{canister_id}/canister.wasm
+    let wasm_dir = get_cache_root()?
+        .join("wasms")
+        .join(canister_id.to_string());
+    let wasm_path = wasm_dir.join("canister.wasm");
+
+    // skip download if cache hit
+    if wasm_path.exists() {
+        let bytes = std::fs::read(&wasm_path)?;
+        let mut hasher = Sha256::new();
+        hasher.update(&bytes);
+        let hash_cache = hasher.finalize();
+
+        if hash_cache.as_slice() == hash_on_chain {
+            info!(logger, "The canister wasm was found in the cache.");
+            return Ok(());
+        }
+    }
+
+    // fetch `dfx:wasm_url`
+    let wasm_url_raw = fetch_metatdata(agent, canister_id, DFX_WASM_URL)
+        .await?
+        .ok_or_else(|| anyhow!("`{DFX_WASM_URL}` metadata not found in canister {canister_id}."))?;
+    let wasm_url_str = String::from_utf8(wasm_url_raw)?;
+    let wasm_url = reqwest::Url::parse(&wasm_url_str)?;
+
+    // download
+    let response = reqwest::get(wasm_url.clone()).await?;
+    let status = response.status();
+    if status.is_client_error() || status.is_server_error() {
+        bail!("Failed to download wasm from url: {wasm_url}.\n  StatusCode: {status}");
+    }
+    let content = response.bytes().await?;
+
+    // hash check
+    let mut hasher = Sha256::new();
+    hasher.update(&content);
+    let hash_download = hasher.finalize();
+    if hash_download.as_slice() != hash_on_chain {
+        bail!(
+            "Hash mismatch.
+  on chain: {}
+  download: {}",
+            hex::encode(hash_on_chain),
+            hex::encode(hash_download.as_slice())
+        );
+    }
+
+    // write to a tempfile
+    let mut f = tempfile::NamedTempFile::new()?;
```

Comment:
This should use `new_in`, to create the temp file on the same filesystem where `wasm_path` will go, so that the `rename` below will take place with an atomic move rather than an interruptible copy.